### PR TITLE
refactor: abstract identify throttling and correct max_concurrency handling

### PR DESCRIPTION
**Pull request by didinele**
**Please describe the changes this PR makes and why it should be merged:**

### Summary
🚀🧪🧹

This pull request improves the sharding and identify strategies for the `ws` package, adding support for cancellation, customization, and throttling of identify requests. It also updates the types, tests, and exports to reflect the new changes. It deletes the obsolete `IdentifyThrottler` module and replaces it with a new `IIdentifyThrottler` interface and a `SimpleIdentifyThrottler` class. It refactors the `WorkerShardingStrategy` and `WorkerContextFetchingStrategy` modules to use the new interface and payload. It fixes a type issue with the `intents` option in the `WebSocketManager` and allows for disabling intents. It adds a new test file for the `SimpleIdentifyThrottler` class and removes an empty line from the `SimpleShardingStrategy` module.

> _`Identify` waits_
> _Throttling and cancelling_
> _Winter of discord_

### Walkthrough
*  Add and implement `IIdentifyThrottler` interface for identify throttling logic ([link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-0e7065c879708496d714e2169454aa085eff268e57c68f88af7458b00c37d86eR1-R11), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-d06653c2251a11ac14f0550d7285a01dce4de801338a6d99015f7b13ece9ad49R1-R50), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-3012d7426356b525d5fb06673ae9c8927af48efe77c8ad6b0e7524921c1304bdL1-R1), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-3012d7426356b525d5fb06673ae9c8927af48efe77c8ad6b0e7524921c1304bdL8-R23), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-3012d7426356b525d5fb06673ae9c8927af48efe77c8ad6b0e7524921c1304bdL35-R34), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L6-R6), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L318-R352), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-73738d0d9c394ee4c1743d566d17da12d1880a5159637b870c2a9bd22950c902L6-R7), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-73738d0d9c394ee4c1743d566d17da12d1880a5159637b870c2a9bd22950c902R32-R35), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-5193ec33d9e2ae076952938af7d77ac7b61cdf51b78a8cc19103f70a599faae2R14), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-5193ec33d9e2ae076952938af7d77ac7b61cdf51b78a8cc19103f70a599faae2R75-R78))
*  Refactor `WorkerShardingStrategy` and `WorkerContextFetchingStrategy` to use new payload structure and abort signals for identify requests ([link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-88a4e3bd1230ef8dbd2f581e319408c8f28a137269423bc9c296190e0a462299L8-R14), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-88a4e3bd1230ef8dbd2f581e319408c8f28a137269423bc9c296190e0a462299L21-R45), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-05d90098e3f175a44ac8e50318b9d9d4fd899fa2c869ff83775ccfa33264926fL12-R22), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-05d90098e3f175a44ac8e50318b9d9d4fd899fa2c869ff83775ccfa33264926fL28-R42), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-05d90098e3f175a44ac8e50318b9d9d4fd899fa2c869ff83775ccfa33264926fL37-R54), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-05d90098e3f175a44ac8e50318b9d9d4fd899fa2c869ff83775ccfa33264926fL49-R93), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L21-R28), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2R42), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L47-R51), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L87-R94), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L125-R131), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L146-R153), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L174-R181), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L190-R197), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L300-R317), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-3853a8134b31283feacc54aefa043b2742de97a58f7db1002eebf9cfecbba27aL120-R120), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-3853a8134b31283feacc54aefa043b2742de97a58f7db1002eebf9cfecbba27aL130-R134), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-3853a8134b31283feacc54aefa043b2742de97a58f7db1002eebf9cfecbba27aL153-R158), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-3853a8134b31283feacc54aefa043b2742de97a58f7db1002eebf9cfecbba27aL171-R173))
*  Update `WebSocketShard` to handle identify request rejection and cancellation ([link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-c8b1fc9d131668594acbc26a02d4033d991968e58df1fb621cb321ee8245a22eL361-R379))
*  Add tests for `SimpleIdentifyThrottler` class using `vitest` framework ([link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-763649ca12adb342d543168677bdb0bed94823aa20f9306442e34eef96cc012bR1-R32))
*  Use type annotations instead of type assertions for consistency and readability in various modules ([link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-0134a3ae3663108422fb01f2cdca4eb587cf5bd4ee158826daa7c147a109d061L60-R62), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-0134a3ae3663108422fb01f2cdca4eb587cf5bd4ee158826daa7c147a109d061L71-R74), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-0134a3ae3663108422fb01f2cdca4eb587cf5bd4ee158826daa7c147a109d061L80-R83), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-0134a3ae3663108422fb01f2cdca4eb587cf5bd4ee158826daa7c147a109d061L90-R103), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-0134a3ae3663108422fb01f2cdca4eb587cf5bd4ee158826daa7c147a109d061L114-R123), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-0134a3ae3663108422fb01f2cdca4eb587cf5bd4ee158826daa7c147a109d061L201-R204),  [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-05d90098e3f175a44ac8e50318b9d9d4fd899fa2c869ff83775ccfa33264926fL37-R54), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L125-R131), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L146-R153), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L174-R181), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L190-R197), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-3853a8134b31283feacc54aefa043b2742de97a58f7db1002eebf9cfecbba27aL130-R134), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-3853a8134b31283feacc54aefa043b2742de97a58f7db1002eebf9cfecbba27aL153-R158), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-3853a8134b31283feacc54aefa043b2742de97a58f7db1002eebf9cfecbba27aL171-R173))
*  Rename `WorkerSendPayloadOp.ShardCanIdentify` to `WorkerSendPayloadOp.ShardIdentifyResponse` and add `ok` property to indicate identify status ([link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-0134a3ae3663108422fb01f2cdca4eb587cf5bd4ee158826daa7c147a109d061L114-R123), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-05d90098e3f175a44ac8e50318b9d9d4fd899fa2c869ff83775ccfa33264926fL28-R42), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L21-R28), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L300-R317), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L318-R352), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-3853a8134b31283feacc54aefa043b2742de97a58f7db1002eebf9cfecbba27aL120-R120))
*  Remove unused `constants` import from `SimpleContextFetchingStrategy` module ([link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-3012d7426356b525d5fb06673ae9c8927af48efe77c8ad6b0e7524921c1304bdL1-R1))
*  Remove unused `throttler` property from `WorkerShardingStrategy` class ([link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-826674b590df2ec66f2ec69cb75d01e1a32d34fc881cb0e8cfaed1f6802baaa2L87-R94))
*  Remove unused `OptionalWebSocketManagerOptions` type from `constants` module ([link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-73738d0d9c394ee4c1743d566d17da12d1880a5159637b870c2a9bd22950c902L6-R7))
*  Remove empty line from `SimpleShardingStrategy` module ([link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-660f7662298fa02bc5feaf40918d7ab8af1b049fa382fad9ce1deb293d9216fdR26))
*  Delete `IdentifyThrottler` class and its test file ([link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-b205755a8ee25528c696f3529d2e6a1dc53cf132858b83446a5b88929fb62b56), [link](https://github.com/discordjs/discord.js/pull/9375/files?diff=unified&w=0#diff-f939c800c26b76caa35370f7ff7a3d8b9ece579d7d965c51dff520ad76a8c26e))

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)
- This PR includes breaking changes (methods removed or renamed, parameters moved or removed)

### CODE REVIEW

**Review by kyranet — APPROVED** [review]
Ref: https://discord.com/developers/docs/topics/gateway#sharding-max-concurrency

How did we miss this 😅

File: packages/ws/src/throttling/IIdentifyThrottler.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+/**
+ * IdentifyThrottlers are responsible for dictating when a shard is allowed to identify.
+ *
+ * See https://discord.com/developers/docs/topics/gateway#sharding-max-concurrency
```
Comment (Jiralite) [inline packages/ws/src/throttling/IIdentifyThrottler.ts:4]:
```suggestion
 * @see {@link https://discord.com/developers/docs/topics/gateway#sharding-max-concurrency}
```

**Review by Jiralite — CHANGES_REQUESTED** [review]

File: packages/ws/src/throttling/SimpleIdentifyThrottler.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+	 */
+	public async waitForIdentify(shardId: number): Promise<void> {
+		const key = shardId % this.maxConcurrency;
+
+		const state = this.states.ensure(key, () => {
+			return {
+				queue: new AsyncQueue(),
+				resetsAt: Number.POSITIVE_INFINITY,
+			};
+		});
```
Comment (kyranet) [inline packages/ws/src/throttling/SimpleIdentifyThrottler.ts:33]:
```suggestion
		const state = this.states.ensure(key, () => ({
			queue: new AsyncQueue(),
			resetsAt: Number.POSITIVE_INFINITY,
		}));
```

**Review by kyranet — CHANGES_REQUESTED** [review]

File: packages/ws/src/throttling/SimpleIdentifyThrottler.ts (hunk 2 above)
Comment (vladfrangu) [reply packages/ws/src/throttling/SimpleIdentifyThrottler.ts:33]: I prefer the current code ngl, but it's up to DD if they wanna move to your suggestion

**Review by vladfrangu — APPROVED** [review]
Code looks good, will try it out and re-review if needed

**Review by kyranet — CHANGES_REQUESTED** [review]

**Review by kyranet — APPROVED** [review]

**Review by SpaceEEC — APPROVED** [review]
Looking good 👀

**Pull request merged**